Fix android detection when python4android is present

tmolitor-stud-tu opened the pull request:
When using **platformdirs** in an android app located on adopted storage, `_android_folder()` in `android.py` returned `None` which lead **platformdirs** to believe that it doesn't run on android and use the `Unix` flavor instead.

In [android.py#L126](https://github.com/platformdirs/platformdirs/blob/main/src/platformdirs/android.py#L126) `autoclass("android.content.Context")` returns a class instead of an instance, which throws an exception on the next line trying to access an instance method named `getFilesDir()`.

That exception is catched and a fallback implementation invoked, that uses a regular expression to search in `sys.path` for an import path rooted in `/data`. Apps installed on adopted storage don't have such paths but begin with `/mnt/expand/` instead.
The regular expression therefore fails and `_android_folder()` returns `None`.

This pull request fixes the issue at least for cases where python4android is available, by first trying to import the `android` module and using it's `mActivity` export to get hold of the main activity. That activity is then used to get a proper application context upon which `getFilesDir()` can be invoked.

If that fails with an exception the original codepath is taken, trying the `autoclass`approach next and if that fails resorting to the regular expression.

gaborbernat reviewed (changes requested):
Test and fix typecheker.

tmolitor-stud-tu commented:
done

gaborbernat reviewed (changes requested):
Please add tests 👍

tmolitor-stud-tu commented:
I've added a test with a mocked android import (`test_android_folder_from_p4a`) and I also added an adopted storage path to `test_android_folder_from_sys_path` and adapted the implementation to do a regex fallback for adopted storage, too.

gaborbernat commented on tests/test_android.py:
```diff
@@ -93,15 +96,42 @@ def test_android_folder_from_jnius(mocker: MockerFixture) -> None:
+
+    _android_folder.cache_clear()
+
+    abspath = MagicMock(return_value="/A")  # pragma: no cover
```
Please remove the no cover pragma and also use snake case notation.

tmolitor-stud-tu replied:
I've just added the no-cover pragma because the `test_android_folder_from_jnius` test had them, too.

Snake case: to be honest I'd rather use the name of the mocked method as variable name, but your linter did not like that, so I went with this stupid sounding `abspath` etc.

I've now changed that to `get_absolute_path` corresponding to the `getAbsolutePath` method. Do you like that more?

gaborbernat replied:
> but your linter did not like that

You are sounding a bit aggressive here, I would suggest adopting a more friendly tone. 

> I've now changed that to `get_absolute_path` corresponding to the `getAbsolutePath` method. Do you like that more?

This is much better and more alongside how things are in Python.

tmolitor-stud-tu replied:
> You are sounding a bit aggressive here, I would suggest adopting a more friendly tone.

Not aggressive, just a bit wearily.

> This is much better and more alongside how things are in Python.

I disagree here. Some projects adopt snail case, some adopt camel case. There is no "this is the only right casing" in python afaic. And well, I often use snail case in python, too. But here I wanted to have the same name for variable and attribute that's mocked.

gaborbernat commented on src/platformdirs/android.py:
```diff
@@ -117,23 +117,50 @@ def site_runtime_dir(self) -> str:
+            result = None
+    if result is None:
+        try:
+            # ...and fall back to using plain pyjnius, if python4android isn't
```
Please realign to 120 characters.

tmolitor-stud-tu replied:
what do you mean? this is less than 120 chars according to my text editor

gaborbernat replied:
You should keep adding words to this line until the next word would go over 120.

tmolitor-stud-tu replied:
That makes this

```
# ...and fall back to using plain pyjnius, if python4android isn't available or doesn't deliver any useful
# result...
```

And I don't think this is particularly readable.
But if you want this, sure, I'll update the pr.

tmolitor-stud-tu replied:
done

gaborbernat reviewed (approved):
![](https://media.tenor.co/images/aa5ae0f203b7e0d862a8d5170db9e52b/tenor.gif)
